### PR TITLE
refactor(manager/circleci): replace regex system with YAML parsing

### PR DESCRIPTION
**Pull request by secustor**
## Changes
Replace regex based parsing with a YAML parser based one including schema. 

## Context
Fixes this issue too: https://github.com/renovatebot/renovate/pull/29902

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/migration.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ export function migrateConfig(config: RenovateConfig): MigratedConfig {
       } else if (is.array(val)) {
         if (is.array(migratedConfig?.[key])) {
           const newArray = [];
-          for (const item of migratedConfig[key] as unknown[]) {
+          for (const item of migratedConfig[key]) {
```
Comment (viceice) [inline lib/config/migration.ts:73]: revert

File: lib/modules/datasource/npm/npmrc.ts [hunk 2]
```diff
@@ -29,7 +29,7 @@ function envReplace(value: any, env = process.env): any {
       logger.warn('Failed to replace env in config: ' + match);
       throw new Error('env-replace');
     }
-    return env[envVarName]!;
+    return env[envVarName];
```
Comment (viceice) [inline lib/modules/datasource/npm/npmrc.ts:32]: revert

File: lib/modules/manager/circleci/__fixtures__/config3.yml [hunk 3]
```diff
@@ -1,18 +1,16 @@
-aliases:  
+aliases:
   - &nodejs
     image: cimg/node:14.8.0
 
 version: 2
 jobs:
-  checkout: 
-    <<: *defaults
```
Comment (viceice) [inline lib/modules/manager/circleci/__fixtures__/config3.yml:8]: why remove?

File: lib/modules/manager/gradle/parser/handlers.ts [hunk 4]
```diff
@@ -371,7 +371,7 @@ export function handleApplyFrom(ctx: Ctx): Ctx {
 
   const matchResult = parseGradle(
     // TODO #22198
-    ctx.fileContents[scriptFilePath]!,
+    ctx.fileContents[scriptFilePath],
```
Comment (viceice) [inline lib/modules/manager/gradle/parser/handlers.ts:374]: revert

File: lib/modules/manager/npm/post-update/npm.ts [hunk 5]
```diff
@@ -234,7 +234,7 @@ export async function generateLockFile(
           if (
             lockFileParsed.packages?.['']?.[depType]?.[lockUpdate.packageName!]
           ) {
-            lockFileParsed.packages[''][depType]![lockUpdate.packageName!] =
+            lockFileParsed.packages[''][depType][lockUpdate.packageName!] =
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/npm.ts:237]: revert

File: lib/modules/manager/nuget/artifacts.ts [hunk 6]
```diff
@@ -163,7 +163,7 @@ export async function updateArtifacts({
           file: {
             type: 'addition',
             path: lockFileName,
-            contents: newLockFileContentMap[lockFileName]!,
+            contents: newLockFileContentMap[lockFileName],
```
Comment (viceice) [inline lib/modules/manager/nuget/artifacts.ts:166]: revert

File: lib/modules/manager/circleci/__fixtures__/config3.yml (hunk 3 above)
Comment (secustor) [reply lib/modules/manager/circleci/__fixtures__/config3.yml:8]: YAML parsing fails because there there is no definition for the `defaults` anchor.

Comment (viceice) [reply lib/modules/manager/circleci/__fixtures__/config3.yml:8]: ok, but is it a valid circleci file? do they provide an implicit default anchor?

we could simply strip such anchors before parsing, like we do in gitlab manager

**Review by viceice — COMMENTED** [review]
see comment

File: lib/modules/manager/circleci/__fixtures__/config3.yml (hunk 3 above)
Comment (secustor) [reply lib/modules/manager/circleci/__fixtures__/config3.yml:8]: I don't thinks so, I can't find any reference to something like default anchor which gets injected into the file before parsing. 

See https://discuss.circleci.com/t/using-defaults-syntax-in-config-yaml-aka-yaml-anchors/16168 for example.

The other test files contain anchors too and parsing fine. 
Actually the fact anchors are now resolved is the reason I had to change the tests.

File: lib/modules/manager/circleci/extract.ts [hunk 7]
```diff
@@ -1,80 +1,50 @@
...
-          },
-          'CircleCI docker image',
-        );
-        dep.depType = 'docker';
-        dep.versioning = 'docker';
-        if (
-          !dep.depName?.startsWith('ubuntu-') &&
-          !dep.depName?.startsWith('windows-server-') &&
-          !dep.depName?.startsWith('android-') &&
-          dep.depName !== 'android'
```
Comment (viceice) [inline lib/modules/manager/circleci/extract.ts:72]: you dropped this condition, it should stay because those values are not resolvable docker images

Comment (secustor) [reply lib/modules/manager/circleci/extract.ts:72]: This on purpose this is not needed anymore. 

Before that it has been used to cover `jobs.release_docker.machine.image` as the regexes looked for `image:`

https://github.com/renovatebot/renovate/blob/34c56e3fe1c4048cfadab6a706553d488c1ef0a0/lib/modules/manager/circleci/__fixtures__/config3.yml#L15-L17

**Review by viceice — APPROVED** [review]

**Pull request merged**